### PR TITLE
Run all examples

### LINKED ISSUE #55: Run all examples on CI
In https://github.com/rusty-fast-solvers/solvers/pull/46, I added a step to the CI to run all the examples in the tools crate. We should update this to run the examples in every crate

### PR DESCRIPTION
**Pull request by mscroggs**
Make the CI run examples from within all crates

Close #55

Added a Python script to search for .rs files in examples directories then create a .sh file that will run them all.

The type of run for each .rs example can be controlled by adding a comment starting with `//?` in the example. eg, the following is included in the examples in tree:

```
//? mpirun -n {{NPROCESSES}} --features "mpi"
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,7 @@ default-members = [
     "quadrature",
     "tools",
     "tree",
+    "hyksort",
```
Comment (mscroggs) [inline Cargo.toml:19]: I'm not sure if hyksort was intentionally left out here, but as it wasn't included, none of its test were running on the CI

**Comment by mscroggs** [comment]
Looks like the examples in the tree crate are failing.

@skailasa do you know if this is because they're out of date or if I'm calling them wrong (do I need to do something special for MPI to work?

**Comment by skailasa** [comment]
> Looks like the examples in the tree crate are failing.
> 
> @skailasa do you know if this is because they're out of date or if I'm calling them wrong (do I need to do something special for MPI to work?

You have to use cargo mpirun which needs to be added to the CI env setup, and specify the feature mpi, as the trees are only optionally compiled with MPI. There's an example of a nice way of doing this and integrating examples as integration tests in the CI in the rsmpi repo.

The command to run the examples is:
cargo mpirun --example test_adaptive -n <nproc> --release --features=mpi

If you don't want to copy the nice workflow in rsmpi just yet that's ok too, however I've put a little stub script in the examples repo that should pick up all the examples and run them as tests, which can probably be edited a little to run from the CI.

**Comment by mscroggs** [comment]
> > Looks like the examples in the tree crate are failing.
> > @skailasa do you know if this is because they're out of date or if I'm calling them wrong (do I need to do something special for MPI to work?
> 
> You have to use cargo mpirun which needs to be added to the CI env setup, and specify the feature mpi, as the trees are only optionally compiled with MPI. There's an example of a nice way of doing this and integrating examples as integration tests in the CI in the rsmpi repo.
> 
> The command to run the examples is: cargo mpirun --example test_x -n nproc --release --features=mpi
> 
> If you don't want to copy the nice workflow in rsmpi just yet that's ok too, however I've put a little stub script in the examples repo that should pick up all the examples and run them as tests, which can probably be edited a little to run from the CI.

Thanks, I'll take a look and try to get this running. (I guess this means I need to do something slightly less simple than `cargo test --examples`)

**Comment by mscroggs** [comment]
> cargo test --lib

The CI is already running this. I'm keen to get it to run the example too though, or they can accidentally go out of date and become a pain to fix when we notice

**Pull request merged**